assumes/expression.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package assumes
+
+import "github.com/juju/version/v2"
+
+var (
+	_ Expression = (*FeatureExpression)(nil)
+	_ Expression = (*CompositeExpression)(nil)
+)
+
+// ExpressionType represents the type of an assumes expression.
+type ExpressionType string
+
+const (
+	AnyOfExpression ExpressionType = "any-of"
+	AllOfExpression ExpressionType = "all-of"
+)
+
+// Expression is an interface implemented by all expression types in this package.
+type Expression interface {
+	Type() ExpressionType
+}
+
+// VersionConstraint describes a constraint for required feature versions.
+type VersionConstraint string
+
+const (
+	VersionGTE VersionConstraint = ">="
+	VersionLT  VersionConstraint = "<"
+)
+
+// FeatureExpression describes a feature that is required by the charm in order
+// to be successfully deployed. Feature expressions may additionally specify a
+// version constraint.
+type FeatureExpression struct {
+	// The name of the feature.
+	Name string
+
+	// A feature within an assumes block may optionally specify a version
+	// constraint.
+	Constraint VersionConstraint
+	Version    *version.Number
+
+	// The raw, unprocessed version string for serialization purposes.
+	rawVersion string
```

Comment:
`version.Number` will expand the version string to a qualified semver value (with `0` for the missing parts). When serializing back we want to preserve the original version as specified by the charm author so we need to record that.

For example, if the assumes expression specifies `1.8` we don't want it to become `1.8.0` because that is not necessarily the same as the version originally specified.

Reply:
Then we shouldn't be using version.Number.